flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationType.scala

```diff
@@ -0,0 +1,171 @@
+    * The provided cumGradient is updated in place.
+    *
+    * @param weightVector The current weight vector
+    * @param lossGradient The vector to which the gradient will be added to, in place.
+    * @return The regularized loss. The gradient is updated in place.
+    */
+  def regularizedLossAndGradient(
+      loss: Double,
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regularizationParameter: Double) : Double ={
+    val adjustedLoss = regLoss(loss, weightVector, regularizationParameter)
+    regGradient(weightVector, lossGradient, regularizationParameter)
+
+    adjustedLoss
+  }
+
+  /** Adds regularization gradient to the loss gradient. The gradient is updated in place **/
+  def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regularizationParameter: Double)
+}
+
+/** Performs no regularization, equivalent to $R(w) = 0$ **/
+class NoRegularization extends RegularizationType {
+  /** Adds regularization to the loss value **/
+  override def regLoss(oldLoss: Double, weightVector: FlinkVector, regularizationParameter: Double):
+  Double = {oldLoss}
+}
+
+/** $L_2$ regularization penalty.
+  *
+  * Penalizes large weights, favoring solutions with more small weights rather than few large ones.
+  *
+  */
+class L2Regularization extends DiffRegularizationType {
+
+  /** Adds regularization to the loss value **/
+  override def regLoss(oldLoss: Double, weightVector: FlinkVector, regParameter: Double)
+    : Double = {
+    val brzVector = weightVector.asBreeze
+    oldLoss + regParameter * (brzVector dot brzVector) / 2
+  }
+
+  /** Adds regularization gradient to the loss gradient. The gradient is updated in place **/
+  override def regGradient(
+      weightVector: FlinkVector,
+      lossGradient: FlinkVector,
+      regParameter: Double): Unit = {
+    BLAS.axpy(regParameter, weightVector, lossGradient)
+  }
+}
+
+/** $L_1$ regularization penalty.
+  *
+  * The $L_1$ penalty can be used to drive a number of the solution coefficients to 0, thereby
+  * producing sparse solutions.
+  *
+  */
+class L1Regularization extends RegularizationType {
+  /** Calculates and applies the regularization amount and the regularization parameter
+    *
+    * Implementation was taken from the Apache Spark Mllib library:
+    * http://git.io/vfZIT
+    * @param oldWeights The old weights
+    * @param effectiveStepSize The effective step size for this iteration
+    * @param regParameter The current regularization parameter
+    * @return A tuple whose first element is the updated weight FlinkVector and the second is the
+    *         regularization value
+    */
+  override def takeStep(
+      oldWeights: FlinkVector,
+      gradient: FlinkVector,
+      effectiveStepSize: Double,
+      regParameter: Double) {
+    BLAS.axpy(-effectiveStepSize, gradient, oldWeights)
+    val brzWeights = oldWeights.asBreeze.toDenseVector
```

Comment:
why converting the oldWeights explicitly to a dense vector. It should be dense since the beginning, or at least we should make sure that it is.